### PR TITLE
Improve ARC handler

### LINKED ISSUE #85: Improve ARC handling
The ARC handler fails when run against randomly generated content:

```
2021-12-02 15:25.12 [info     ] Calculating chunk for YARA match identifier=$arc_magic real_offset=0x554816 start_offset=0x554816
2021-12-02 15:25.12 [debug    ] Header parsed                  header=
00000000  1a 03 1b da e2 92 33 be  29 e5 72 b0 7b 71 fe c8   ......3.).r.{q..
00000010  80 11 57 34 68 96 3e 58  2a 9a 0a cc 26            ..W4h.>X*...&

struct heads:
- archive_marker: 0x1a
- header_type: 0x3
- name: b'\x1b\xda\xe2\x923\xbe)\xe5r\xb0{q\xfe'
- size: 0x571180c8
- date: 0x6834
- time: 0x3e96
- crc: 0x2a58
- length: 0x26cc0a9a
2021-12-02 15:25.12 [error    ] Unhandled Exception during chunk calculation 
Traceback (most recent call last):
  File "/home/quentin/unblob/unblob/strategies.py", line 50, in search_chunks_by_priority
    chunk = handler.calculate_chunk(limited_reader, real_offset)
  File "/home/quentin/unblob/unblob/handlers/archive/arc.py", line 67, in calculate_chunk
    header = self.parse_header(file)
  File "/home/quentin/unblob/unblob/models.py", line 119, in parse_header
    header = self._struct_parser.parse(self.HEADER_STRUCT, file, endian)
  File "/home/quentin/unblob/unblob/file_utils.py", line 143, in parse
    return struct_parser(file)
  File "/home/quentin/.cache/pypoetry/virtualenvs/unblob-U71Ryf-L-py3.8/lib/python3.8/site-packages/dissect/cstruct/types/base.py", line 16, in __call__
    return self.read(*args, **kwargs)
  File "/home/quentin/.cache/pypoetry/virtualenvs/unblob-U71Ryf-L-py3.8/lib/python3.8/site-packages/dissect/cstruct/types/base.py", line 62, in read
    return self._read(obj)
  File "<compiled heads>", line 14, in _read
EOFError
```

We should capture the EOFError early in the handler and simply not return a chunk.

### PR DESCRIPTION
**Pull request by QKaiser**
Fix #85 by catching errors triggered when we seek past the file.

### CODE REVIEW

File: unblob/handlers/archive/arc.py [hunk 1]
```diff
@@ -50,22 +50,26 @@ def calculate_chunk(
...
+
+                if read_bytes == END_HEADER:
+                    offset += 2
+                    break
+                file.seek(offset)
+                header = self.parse_header(file)
+                offset += len(header) + header.size
+
+        except (EOFError, OSError):
+            # OSError triggered when seeking past the EOF
```
Comment (kukovecz) [inline unblob/handlers/archive/arc.py:68]: Can you show an example? I believe this is not true. We just talked about this with @kissgyorgy, that `seek` does not trigger any error in this case.

His example to prove:
```python
>>> from pathlib import Path
>>> path = Path("pyproject.toml")
>>> path.stat().st_size
744
>>> fp = path.open("rb")
>>> fp.seek(1000)
1000
>>> fp.tell()
1000
>>> fp.read()
b''
```

Comment (vlaci) [reply unblob/handlers/archive/arc.py:68]: >       lseek() allows the file offset to be set beyond the end of the
>       file (but this does not change the size of the file).

**Comment by QKaiser** [comment]
Fixed and rebased !

File: unblob/handlers/archive/arc.py [hunk 2]
```diff
@@ -50,22 +50,26 @@ def calculate_chunk(
...
+
+                if read_bytes == END_HEADER:
+                    offset += 2
+                    break
+                file.seek(offset)
+                header = self.parse_header(file)
+                offset += len(header) + header.size
+
+        except (EOFError):
+            # OSError triggered when seeking past the EOF
```
Comment (vlaci) [inline unblob/handlers/archive/arc.py:68]: On a related note: what throws the `EOFError`? You should always be able to read from an open file. After some digging I found that cstruct can raise it. I'd adjust this comment accordingly
```suggestion
        except EOFError:
            # EOFError triggered when file ends during header parsing
```

**Review by vlaci — CHANGES_REQUESTED** [review]

File: unblob/handlers/archive/arc.py (hunk 2 above)
Comment (kissgyorgy) [reply unblob/handlers/archive/arc.py:68]: If you put the try: except: block around the self.parse_header call, you don't need the comment. EOFError can't be raised from other parts of this code.

File: unblob/handlers/archive/arc.py (hunk 1 above)
Comment (QKaiser) [reply unblob/handlers/archive/arc.py:68]: Finally found what triggered the OSError: negative seeking !

```
>>> with open("/tmp/garbage.bin", "rb") as f:
...     f.seek(-1)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
OSError: [Errno 22] Invalid argument
```

Comment (QKaiser) [reply unblob/handlers/archive/arc.py]: Fixed this by switching the ARC header size from `long `to `ulong`. Having a negative size is not right.

File: unblob/handlers/archive/arc.py (hunk 2 above)
Comment (QKaiser) [reply unblob/handlers/archive/arc.py]: Got rid of the comment and placed the try/except block around `parse_header` :)

File: unblob/handlers/archive/arc.py [hunk 3]
```diff
@@ -54,17 +54,19 @@ def calculate_chunk(
...
+            read_bytes = file.read(2)
 
             if read_bytes == END_HEADER:
                 offset += 2
                 break
             file.seek(offset)
-            header = self.parse_header(file)
+            try:
+                header = self.parse_header(file)
+            except (EOFError):
```
Comment (vlaci) [inline unblob/handlers/archive/arc.py:65]:
```suggestion
            except EOFError:
```

Comment (QKaiser) [reply unblob/handlers/archive/arc.py]: Done.

**Review by kissgyorgy — APPROVED** [review]
We fixed the comments in pair programming, the whole team was there.

**Pull request merged**